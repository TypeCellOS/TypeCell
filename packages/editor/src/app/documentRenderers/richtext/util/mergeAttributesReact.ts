import { mergeAttributes } from "@tiptap/core";

export default function mergeAttributesReact(
  ...objects: Record<string, any>[]
): Record<string, any> {
  const ret = mergeAttributes(...objects);
  if (ret.class) {
    // react uses className instead of class
    ret.className = ret.class;
    delete ret.class;
  }
  return ret;
}
