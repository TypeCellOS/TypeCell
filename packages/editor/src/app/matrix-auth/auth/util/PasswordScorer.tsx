/*
Copyright 2018 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import zxcvbn from "zxcvbn";

const ZXCVBN_USER_INPUTS = ["riot", "matrix", "typecell"];

/**
 * Wrapper around zxcvbn password strength estimation
 * Include this only from async components: it pulls in zxcvbn
 * (obviously) which is large.
 *
 * @param {string} password Password to score
 * @returns {object} Score result with `score` and `feedback` properties
 */
export function scorePassword(password: string) {
  if (password.length === 0) return null;

  const userInputs = ZXCVBN_USER_INPUTS.slice();
  //   if (MatrixClientPeg.get()) {
  // userInputs.push(MatrixClientPeg.get().getUserIdLocalpart());
  //   }

  let zxcvbnResult = zxcvbn(password, userInputs);
  // Work around https://github.com/dropbox/zxcvbn/issues/216
  if (password.includes(" ")) {
    const resultNoSpaces = zxcvbn(password.replace(/ /g, ""), userInputs);
    if (resultNoSpaces.score < zxcvbnResult.score)
      zxcvbnResult = resultNoSpaces;
  }

  for (let i = 0; i < zxcvbnResult.feedback.suggestions.length; ++i) {
    // translate suggestions
    zxcvbnResult.feedback.suggestions[i] = zxcvbnResult.feedback.suggestions[i];
  }
  // and warning, if any
  if (zxcvbnResult.feedback.warning) {
    zxcvbnResult.feedback.warning = zxcvbnResult.feedback.warning;
  }

  return zxcvbnResult;
}
