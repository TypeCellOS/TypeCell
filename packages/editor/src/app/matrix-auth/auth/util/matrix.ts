export interface IMatrixClientCreds {
  homeserverUrl: string;
  identityServerUrl: string | undefined;
  userId: string;
  deviceId?: string;
  accessToken: string;
  guest?: boolean;
  pickleKey?: string | undefined;
  freshLogin?: boolean;
}
