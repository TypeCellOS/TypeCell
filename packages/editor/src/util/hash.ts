export function hash(str: string) {
  var hash = 0;
  for (var i = 0; i < str.length; i++) {
    var character = str.charCodeAt(i);
    hash = (hash << 5) - hash + character;
    hash = hash & hash; // Convert to 32bit integer
  }
  return hash;
}
