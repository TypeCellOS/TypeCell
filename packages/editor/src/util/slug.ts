import speakingurl from "speakingurl";

export function slug(str: string) {
  return speakingurl(str, {
    custom: {
      "@": "@", // TODO: necesary?
    },
  });
}
