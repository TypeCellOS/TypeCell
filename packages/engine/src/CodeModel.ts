type Event<T> = {
  (listener: (e: T) => any): { dispose: () => void };
};

export type CodeModel = {
  onWillDispose: Event<void>;
  onDidChangeContent: Event<void>;
  getValue(): string;
  readonly path: string;
  readonly language: string;
};
