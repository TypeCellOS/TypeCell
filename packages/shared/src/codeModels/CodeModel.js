export {};
//# sourceMappingURL=CodeModel.js.map