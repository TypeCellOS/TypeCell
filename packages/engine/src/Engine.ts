import { lifecycle, event } from "vscode-lib";
import { createCellEvaluator } from "./CellEvaluator";
import { CodeModel } from "./CodeModel";
import { createContext, TypeCellContext } from "./context";

export type ResolvedImport = {
  module: any;
} & lifecycle.IDisposable;

/**
 * The engine automatically runs models registered to it.
 * The code of the models is passed a context ($) provided by the engine.
 * This context is how the code of different models can react to each other.
 *
 * @export
 * @class Engine
 */
export class Engine<T extends CodeModel> extends lifecycle.Disposable {
  private disposed = false;
  public readonly observableContext = createContext<any>({} as any);
  private readonly registeredModels = new Set<T>();

  private readonly evaluatorCache = new Map<
    T,
    ReturnType<typeof createCellEvaluator>
  >();

  private readonly _onOutput: event.Emitter<{ model: T; output: any }> =
    this._register(new event.Emitter<{ model: T; output: any }>());

  /**
   * Raised whenever a model is (re)evaluated, with the exports by that model
   *
   * @type {event.Event<{ model: T; output: any }>}
   * @memberof Engine
   */
  public readonly onOutput: event.Event<{ model: T; output: any }> =
    this._onOutput.event;

  private readonly _onBeforeExecution: event.Emitter<{ model: T }> =
    this._register(new event.Emitter<{ model: T }>());

  /**
   * Raised before (re)evaluating a model
   *
   * @type {event.Event<{ model: T }>}
   * @memberof Engine
   */
  public readonly onBeforeExecution: event.Event<{ model: T }> =
    this._onBeforeExecution.event;

  /**
   * @param resolveImport Function (resolver) to resolve imports
   */
  constructor(
    private resolveImport: (
      module: string,
      forModel: T
    ) => Promise<ResolvedImport>
  ) {
    super();
  }

  public registerModelProvider(provider: {
    onDidCreateCompiledModel: event.Event<T>;
    compiledModels: T[];
  }) {
    provider.compiledModels.forEach(this.registerModel);
    this._register(
      provider.onDidCreateCompiledModel((m) => {
        this.registerModel(m);
      })
    );
  }

  /**
   * Register a model to the engine. After registering, the model will be observed for changes and automatically re-evaluated.
   *
   * When the model is disposed (model.dispose()), the model is automatically unregistered.
   * @param model model to register
   */
  public registerModel(model: T) {
    if (this.disposed) {
      throw new Error("registering model on disposed engine");
    }
    if (this.registeredModels.has(model)) {
      console.warn("model already registered"); // TODO: shouldn't happen
      return;
    }
    this.registeredModels.add(model);
    const evaluate = () => {
      this.evaluateUpdate(
        model,
        this.observableContext,
        async (moduleName: string) => {
          const ret = await this.resolveImport(moduleName, model);
          this._register(ret);
          if (this.disposed) {
            // disposed in the meantime
            ret.dispose();
          }
          return ret.module;
        },
        (model, output) => this._onOutput.fire({ model, output })
      ); // catch errors?
    };
    let prevValue: string | undefined = model.getValue();

    this._register(
      model.onDidChangeContent((_event) => {
        if (model.getValue() !== prevValue) {
          // make sure there were actual changes from the previous value

          prevValue = model.getValue();
          evaluate();
        } else {
          // TODO: inspect when this is the case. For initialization it makes sense,
          // but why do we get duplicate events more often?
        }
      })
    );

    // evaluate initial
    evaluate();

    this._register(
      model.onWillDispose(() => {
        this.registeredModels.delete(model);
        const evaluator = this.evaluatorCache.get(model);
        if (evaluator) {
          evaluator.dispose();
          this.evaluatorCache.delete(model);
        }
      })
    );
  }

  public dispose() {
    if (this.disposed) {
      throw new Error("Engine already disposed");
    }
    this.disposed = true;

    super.dispose();
    this.evaluatorCache.forEach((e) => e.dispose());
  }

  private async evaluateUpdate(
    model: T,
    typecellContext: TypeCellContext<any>,
    resolveImport: (module: string) => Promise<any>,
    onOutput: (model: T, output: any) => void
  ) {
    if (!this.evaluatorCache.has(model)) {
      this.evaluatorCache.set(
        model,
        createCellEvaluator(
          typecellContext,
          resolveImport,
          true,
          (output) => onOutput(model, output),
          () => this._onBeforeExecution.fire({ model })
        )
      );
    }
    const evaluator = this.evaluatorCache.get(model)!;
    if (model.language !== "javascript") {
      throw new Error("can not evaluate non-javascript code");
    }
    let code = model.getValue();
    await evaluator.evaluate(code);
  }
}
