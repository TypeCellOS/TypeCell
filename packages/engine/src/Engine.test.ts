import { CodeModel } from "./CodeModel";
import { Engine } from "./Engine";
import { event } from "vscode-lib";
import {
  buildMockedModel,
  importResolver,
  toAMDFormat,
  waitTillEvent,
} from "./tests/util/helpers";
import { CodeModelMock } from "./tests/util/CodeModelMock";

describe("engine class execution", function () {
  describe("basic model execution", () => {
    const getModel1 = () =>
      buildMockedModel(
        "model1",
        `let x = 4; 
let y = 6; 
let sum = x + y;
exports.sum = sum;
exports.default = sum;`
      );

    const getModel2 = () =>
      buildMockedModel("model2", `exports.default = $.sum - 5;`);

    it("should execute a single model", async () => {
      const engine = new Engine<CodeModel>(importResolver);
      engine.registerModel(getModel1());

      const { model, output } = await event.Event.toPromise(engine.onOutput);

      expect(model.path).toBe("model1");
      expect(output.sum).toBe(10);
      expect(output.default).toBe(10);
    });

    it("should read exported variables from other models", async () => {
      const engine = new Engine<CodeModel>(importResolver);
      engine.registerModel(getModel1());
      await event.Event.toPromise(engine.onOutput);

      engine.registerModel(getModel2());
      const { model, output } = await event.Event.toPromise(engine.onOutput);

      expect(model.path).toBe("model2");
      expect(output.default).toBe(5);
    });

    it("should re-evaluate code after change", async () => {
      const engine = new Engine<CodeModel>(importResolver);
      const model1 = getModel1();

      engine.registerModel(model1);
      await event.Event.toPromise(engine.onOutput);

      model1.updateCode(
        toAMDFormat(`let x = 0;
    let y = 6;
    let sum = x + y;
    exports.sum = sum;
    exports.default = sum;`)
      );

      const { output } = await event.Event.toPromise(engine.onOutput);

      expect(output.sum).toBe(6);
      expect(output.default).toBe(6);
    });

    it("should re-evaluate other models when global variable changes", async () => {
      const engine = new Engine<CodeModel>(importResolver);
      // TODO: Expected 4 events. Figure out why model 2 re-evaluates.
      const eventsPromise = waitTillEvent(engine.onOutput, 5);
      const model1 = getModel1();
      const model2 = getModel2();

      engine.registerModel(model1);
      engine.registerModel(model2);

      model1.updateCode(
        toAMDFormat(`let x = 0;
    let y = 6;
    let sum = x + y;
    exports.sum = sum;
    exports.default = sum;`)
      );

      const events = await eventsPromise;
      const eventsSnapshot = events.map((event) => ({
        path: event.model.path,
        output: event.output,
      }));
      const finalEvent = eventsSnapshot[eventsSnapshot.length - 1];

      expect(finalEvent.path).toBe("model2");
      expect(finalEvent.output.default).toBe(1);
      expect(eventsSnapshot).toMatchSnapshot();
    });
  });

  describe("console messages", () => {
    const getModel1 = () => buildMockedModel("model1", `console.log('hi!');`);
    const getModel2 = () =>
      buildMockedModel(
        "model2",
        `console.info('info'); console.warn('warn'); console.error('error');`
      );
    const getModel3 = () =>
      buildMockedModel(
        "model3",
        `console.log('before');
        await new Promise((resolve)=> {
          setTimeout(()=> {
            resolve();
          }, 1)
        });
        console.log('after');`
      );
    const getModel4 = () =>
      new CodeModelMock(
        "javascript",
        "model4",
        `define(["require", "exports", "logdown"], function(require, exports, logdown) {
          "use strict";
          Object.defineProperty(exports, "__esModule", { value: true });         
        
          let logger = logdown("logger 1");
          logger.state.isEnabled = true;
          logger.log("message 1");

          setTimeout(() => {
            logger.state.isEnabled = true;
            logger.log("message 2");
          }, 1);
        });`
      );

    it("should capture console.log message", async () => {
      const engine = new Engine<CodeModel>(importResolver);
      const eventsPromise = waitTillEvent(engine.onConsole, 1);
      const model1 = getModel1();

      engine.registerModel(model1);

      const consoleEvents = await eventsPromise;

      expect(consoleEvents[0].payload.method).toBe("log");
      expect(consoleEvents[0].payload.arguments[0]).toBe("hi!");
    });

    it("should capture console.warn/info/error messages", async () => {
      const engine = new Engine<CodeModel>(importResolver);
      const eventsPromise = waitTillEvent(engine.onConsole, 3);
      const model2 = getModel2();

      engine.registerModel(model2);

      const events = await eventsPromise;
      const eventsSnapshot = events.map((event) => {
        return {
          path: event.model.path,
          console: event.payload,
        };
      });
      expect(eventsSnapshot).toMatchSnapshot();
    });

    it("should capture console.log messages after async", async () => {
      const engine = new Engine<CodeModel>(importResolver);
      const eventsPromise = waitTillEvent(engine.onConsole, 2);
      const model3 = getModel3();

      engine.registerModel(model3);

      const events = await eventsPromise;
      expect(events[0].payload.arguments[0]).toBe("before");
      expect(events[1].payload.arguments[0]).toBe("after");
    });

    it("should capture console.log messages from library (sync only)", async () => {
      const engine = new Engine<CodeModel>(importResolver);
      const eventsPromise = waitTillEvent(engine.onConsole, 2);
      const model4 = getModel4();

      engine.registerModel(model4);

      const events = await eventsPromise;
      expect(events[0].payload.arguments[1]).toBe("message 1");
      expect(events[1].payload.arguments[1]).toBe("message 2");
    });
  });
});
