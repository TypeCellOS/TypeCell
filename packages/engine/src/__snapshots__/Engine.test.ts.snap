// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engine class execution basic model execution should re-evaluate other models when global variable changes 1`] = `
Array [
  Object {
    "output": Object {
      "default": 10,
    },
    "path": "model1",
  },
  Object {
    "output": Object {
      "default": 5,
    },
    "path": "model2",
  },
  Object {
    "output": Object {
      "default": 5,
    },
    "path": "model2",
  },
  Object {
    "output": Object {
      "default": 6,
    },
    "path": "model1",
  },
  Object {
    "output": Object {
      "default": 1,
    },
    "path": "model2",
  },
]
`;

exports[`engine class execution console messages should capture console.warn/info/error messages 1`] = `
Array [
  Object {
    "console": Object {
      "arguments": Array [
        "info",
      ],
      "method": "info",
    },
    "path": "model2",
  },
  Object {
    "console": Object {
      "arguments": Array [
        "warn",
      ],
      "method": "warn",
    },
    "path": "model2",
  },
  Object {
    "console": Object {
      "arguments": Array [
        "error",
      ],
      "method": "error",
    },
    "path": "model2",
  },
]
`;
